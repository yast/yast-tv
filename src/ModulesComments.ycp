/**
 * File:	modules/ModulesComments.ycp
 * Package:	TV cards configuration
 * Summary:	Library for handling special comments in /etc/modules.conf
 * Authors:	Jan Holesovsky <kendy@suse.cz>
 *
 * $Id$
 *
 * Functions for handling special comments in /etc/modules.conf
 */

{

    module "ModulesComments";

    /**
     * Create comment for modules.conf from the unique key and the name
     * of the card. It will be parsed when reading.
     * This comment should be placed before the alias differentiating
     * the card (char-major-81-x for TV cards, ethx for ethernet, ...)
     * @param name Name of the card
     * @param unique_key Unique key
     * @return string The comment.
     * @example
     *    ModulesComments::StoreToComment ("Ultra brutal TV card", "xyza.aiLKJkjsdlj")
     * -> "# xyza.aiLKJkjsdlj:Ultra brutal TV card\n"
     */
    global define string StoreToComment (string name, string unique_key) ``{
	if (name == nil)
	{
	    name = "Unknown card";
	}
	if (unique_key == nil)
	{
	    unique_key = "uniq.key";
	}

	return sformat ( "# %1:%2\n", unique_key, name );
    }

    /**
     * Extacts the unique key and the name of the card from comment in the
     * modules.conf placed before the alias differentiatin the card.
     * @param comment The comment
     * @return map Returns map: $[ "unique_key" : string, "name" : string ]
     * @example
     *    ModulesComments::ExtractFromComment ("# xyza.aiLKJkjsdlj:Ultra brutal TV card\n")
     * -> $[ "name" : "Ultra brutal TV card", "unique_key" : "xyza.aiLKJkjsdlj" ]
     */
    global define map ExtractFromComment(string comment) ``{
	if (comment == nil)
	{
	    comment = "";
	}

	map result = $[ "name" : "", "unique_key" : "" ];

	// split to lines
	list comment_lines = splitstring(comment, "\n");

	// find last line with a ":"
	string line_with_uk = "";
	integer colon_pos_uk = nil;
	foreach(`line, comment_lines, ``{
	    integer c_pos = findfirstof(line, ":");
	    if (c_pos != nil)
	    {
		line_with_uk = line;
		colon_pos_uk = c_pos;
	    }
	});

	// did we find it?
	if (colon_pos_uk != nil)
	{
	    // extract name
	    string name = substring(line_with_uk,
				    colon_pos_uk + 1,
				    size(line_with_uk) - colon_pos_uk - 1);
	    // extract unique key
	    integer start_uk = findfirstnotof(line_with_uk, "# \t");
	    string uk = substring(line_with_uk,
				  start_uk,
				  colon_pos_uk - start_uk);

	    result = $[ "name" : name, "unique_key" : uk ];
	}

	return result;
    }

/* EOF */
}
