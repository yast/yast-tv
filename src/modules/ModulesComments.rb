# encoding: utf-8

# File:	modules/ModulesComments.ycp
# Package:	TV cards configuration
# Summary:	Library for handling special comments in /etc/modules.conf
# Authors:	Jan Holesovsky <kendy@suse.cz>
#
# $Id$
#
# Functions for handling special comments in /etc/modules.conf
require "yast"

module Yast
  class ModulesCommentsClass < Module
    # Create comment for modules.conf from the unique key and the name
    # of the card. It will be parsed when reading.
    # This comment should be placed before the alias differentiating
    # the card (char-major-81-x for TV cards, ethx for ethernet, ...)
    # @param [String] name Name of the card
    # @param [String] unique_key Unique key
    # @return [String] The comment.
    # @example
    #    ModulesComments::StoreToComment ("Ultra brutal TV card", "xyza.aiLKJkjsdlj")
    # -> "# xyza.aiLKJkjsdlj:Ultra brutal TV card\n"
    def StoreToComment(name, unique_key)
      name = "Unknown card" if name == nil
      unique_key = "uniq.key" if unique_key == nil

      Builtins.sformat("# %1:%2\n", unique_key, name)
    end

    # Extacts the unique key and the name of the card from comment in the
    # modules.conf placed before the alias differentiatin the card.
    # @param [String] comment The comment
    # @return [Hash] Returns map: $[ "unique_key" : string, "name" : string ]
    # @example
    #    ModulesComments::ExtractFromComment ("# xyza.aiLKJkjsdlj:Ultra brutal TV card\n")
    # -> $[ "name" : "Ultra brutal TV card", "unique_key" : "xyza.aiLKJkjsdlj" ]
    def ExtractFromComment(comment)
      comment = "" if comment == nil

      result = { "name" => "", "unique_key" => "" }

      # split to lines
      comment_lines = Builtins.splitstring(comment, "\n")

      # find last line with a ":"
      line_with_uk = ""
      colon_pos_uk = nil
      Builtins.foreach(
        Convert.convert(comment_lines, :from => "list", :to => "list <string>")
      ) do |line|
        c_pos = Builtins.findfirstof(line, ":")
        if c_pos != nil
          line_with_uk = line
          colon_pos_uk = c_pos
        end
      end

      # did we find it?
      if colon_pos_uk != nil
        # extract name
        name = Builtins.substring(
          line_with_uk,
          Ops.add(colon_pos_uk, 1),
          Ops.subtract(
            Ops.subtract(Builtins.size(line_with_uk), colon_pos_uk),
            1
          )
        )
        # extract unique key
        start_uk = Builtins.findfirstnotof(line_with_uk, "# \t")
        uk = Builtins.substring(
          line_with_uk,
          start_uk,
          Ops.subtract(colon_pos_uk, start_uk)
        )

        result = { "name" => name, "unique_key" => uk }
      end

      deep_copy(result)
    end

    publish :function => :StoreToComment, :type => "string (string, string)"
    publish :function => :ExtractFromComment, :type => "map (string)"
  end

  ModulesComments = ModulesCommentsClass.new
end
