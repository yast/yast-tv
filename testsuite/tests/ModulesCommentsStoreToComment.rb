# encoding: utf-8

module Yast
  class ModulesCommentsStoreToCommentClient < Client
    def main
      Yast.import "ModulesComments"
      Yast.include self, "testsuite.rb"

      DUMP([ModulesComments.StoreToComment(nil, nil)])

      DUMP([ModulesComments.StoreToComment("Test TV Card", nil)])

      DUMP([ModulesComments.StoreToComment(nil, "dn0t.+xOL8ZCSAQC")])

      DUMP([ModulesComments.StoreToComment("My TV Card", "nd54sdde8d25cv45")])

      nil
    end
  end
end

Yast::ModulesCommentsStoreToCommentClient.new.main
