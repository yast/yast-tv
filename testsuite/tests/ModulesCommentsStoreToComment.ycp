{
    import "ModulesComments";
    include "testsuite.ycp";

    DUMP([ ModulesComments::StoreToComment (nil, nil) ]);

    DUMP([ ModulesComments::StoreToComment ("Test TV Card", nil) ]);

    DUMP([ ModulesComments::StoreToComment (nil, "dn0t.+xOL8ZCSAQC") ]);

    DUMP([ ModulesComments::StoreToComment ("My TV Card", "nd54sdde8d25cv45") ]);
}
