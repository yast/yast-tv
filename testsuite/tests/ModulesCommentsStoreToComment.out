Dump	["# uniq.key:Unknown card\n"]
Dump	["# uniq.key:Test TV Card\n"]
Dump	["# dn0t.+xOL8ZCSAQC:Unknown card\n"]
Dump	["# nd54sdde8d25cv45:My TV Card\n"]
